spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/SparkV2Filters.java

```diff
@@ -0,0 +1,385 @@
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value()));
+          }
+
+        case EQ: // used for both eq and null-safe-eq
+          Object value;
+          String attributeName;
+          if (predicate.children()[SECOND_ORDINAL] instanceof LiteralValue) {
+            attributeName = predicate.children()[FIRST_ORDINAL].toString();
+            value = convertUTF8StringIfNecessary(((LiteralValue) predicate.children()[1]).value());
+          } else {
+            attributeName = predicate.children()[SECOND_ORDINAL].toString();
+            value =
+                convertUTF8StringIfNecessary(
+                    ((LiteralValue) predicate.children()[FIRST_ORDINAL]).value());
+          }
+
+          if (predicate.name().equals("=")) {
+            // comparison with null in normal equality is always null. this is probably a mistake.
+            Preconditions.checkNotNull(
+                value, "Expression is always false (eq is not null-safe): %s", predicate);
+            return handleEqual(unquote(attributeName), value);
+          } else { // "<=>"
+            if (value == null) {
+              return isNull(unquote(attributeName));
+            } else {
+              return handleEqual(unquote(attributeName), value);
+            }
+          }
+
+        case IN:
+          return in(
+              unquote(predicate.children()[FIRST_ORDINAL].toString()),
+              Arrays.stream(predicate.children())
+                  .skip(1)
+                  .map(val -> convertUTF8StringIfNecessary(((LiteralValue) val).value()))
+                  .filter(Objects::nonNull)
+                  .collect(Collectors.toList()));
+
+        case NOT:
+          Not notFilter = (Not) predicate;
+          Predicate childFilter = notFilter.child();
+          if (childFilter.name().equals("IN")) {
+            // infer an extra notNull predicate for Spark NOT IN filters
+            // as Iceberg expressions don't follow the 3-value SQL boolean logic
+            // col NOT IN (1, 2) in Spark is equivalent to notNull(col) && notIn(col, 1, 2) in
+            // Iceberg
+            Expression notIn =
+                notIn(
+                    unquote(childFilter.children()[FIRST_ORDINAL].toString()),
+                    Arrays.stream(childFilter.children())
+                        .skip(1)
+                        .map(val -> convertUTF8StringIfNecessary(((LiteralValue) val).value()))
+                        .filter(Objects::nonNull)
+                        .collect(Collectors.toList()));
+            return and(notNull(unquote(childFilter.children()[FIRST_ORDINAL].toString())), notIn);
+          } else if (hasNoInFilter(childFilter)) {
+            Expression child = convert(childFilter);
+            if (child != null) {
+              return not(child);
+            }
+          }
+          return null;
+
+        case AND:
+          {
```

Comment:
nit: Do we need these extra `{}` here and 2 branches below?

Reply:
This is required. Otherwise, I can't use the same variable names (`left` and `right`) in difference case blocks.